### PR TITLE
feat(utxo-lib): create p2trMusig2 nonce using derived key

### PR DESCRIPTION
**Pull request by saravanan7mani**
Add a function to create p2trMusig2 nonce directly using derived key. It is done to make it consistent with other sign functions.

TICKET: BG-73060

### CODE REVIEW

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 1]
```diff
@@ -391,22 +391,30 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
-    } else if (input.bip32Derivation && input.bip32Derivation.length > 0) {
-      return UtxoPsbt.deriveWalletKey(bip32, input.bip32Derivation, { keyFormat: 'plain' })?.publicKey;
+    if ((!inputType || inputType === 'taproot') && input.tapBip32Derivation?.length) {
+      return UtxoPsbt.deriveWalletKey(bip32, input.tapBip32Derivation, { keyFormat: 'xOnly' });
     }
-    return bip32.publicKey;
+    if ((!inputType || inputType === 'non-taproot') && input.bip32Derivation?.length) {
+      return UtxoPsbt.deriveWalletKey(bip32, input.bip32Derivation, { keyFormat: 'plain' });
+    }
+    return bip32;
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:417]:
```suggestion
    return undefined;
```

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 2]
```diff
@@ -955,14 +964,14 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
     }
-    if (!input.tapBip32Derivation?.length) {
+    if (keyType === 'root' && !input.tapBip32Derivation?.length) {
       throw new Error('tapBip32Derivation is required to create nonce');
     }
-    const derivedWalletKey = UtxoPsbt.deriveWalletKey(rootWalletKey, input.tapBip32Derivation, { keyFormat: 'xOnly' });
-    if (!derivedWalletKey) {
-      throw new Error('No bip32Derivation masterFingerprint matched the rootWalletKey fingerprint');
+    const signerKeyPair = keyType === 'root' ? this.getSignerKeyPairForInput(inputIndex, keyPair, 'taproot') : keyPair;
+    if (!signerKeyPair) {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:971]:
```suggestion
    const derivedWalletKey = keyType === 'root' ? this.getSignerKeyPairForInput(inputIndex, keyPair, 'taproot') : keyPair;
    if (!derivedWalletKey) {
```

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 3]
```diff
@@ -955,14 +964,14 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+      throw new Error('No bip32Derivation masterFingerprint matched the HD keyPair fingerprint');
     }
-    if (!derivedWalletKey.privateKey) {
+    if (!signerKeyPair.privateKey) {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:974]:
```suggestion
    if (!derivedKeyPair.privateKey) {
```

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 4]
```diff
@@ -971,15 +980,15 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
     }
     assertPsbtMusig2Participants(participants, input.tapInternalKey, input.tapMerkleRoot);
     const { tapOutputKey, participantPubKeys } = participants;
-    const participantPubKey = participantPubKeys.find((pubKey) => pubKey.equals(derivedWalletKey.publicKey));
+    const participantPubKey = participantPubKeys.find((pubKey) => pubKey.equals(signerKeyPair.publicKey));
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:983]:
```suggestion
    const participantPubKey = participantPubKeys.find((pubKey) => pubKey.equals(derivedKeyPair.publicKey));
```

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 5]
```diff
@@ -1011,11 +1010,39 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+   * Generates and sets Musig2 nonces to p2trMusig2 key path spending inputs.
+   * tapInternalkey, tapMerkleRoot, tapBip32Derivation for derivedWalletKey are required per p2trMusig2 key path input.
+   * Also participant keys are required from psbt proprietary key values.
+   * Ref: https://gist.github.com/sanket1729/4b525c6049f4d9e034d27368c49f28a6
+   * @param psbt
+   * @param derivedKeyPair derived key pair
+   * @param sessionId Optional extra entropy. If provided it must either be a counter unique to this secret key,
+   * (converted to an array of 32 bytes), or 32 uniformly random bytes.
+   */
+  setMusig2Nonces(derivedKeyPair: BIP32Interface, sessionId?: Buffer): void {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:1028]:
```suggestion
  setMusig2Nonces(derivedKeyPair: BIP32Interface, sessionId?: Buffer): this {
```

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 6]
```diff
@@ -1011,11 +1010,39 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+   * Generates and sets Musig2 nonces to p2trMusig2 key path spending inputs.
+   * tapInternalkey, tapMerkleRoot, tapBip32Derivation for derivedWalletKey are required per p2trMusig2 key path input.
+   * Also participant keys are required from psbt proprietary key values.
+   * Ref: https://gist.github.com/sanket1729/4b525c6049f4d9e034d27368c49f28a6
+   * @param psbt
+   * @param hdKeyPair HD key pair
+   * @param sessionId Optional extra entropy. If provided it must either be a counter unique to this secret key,
+   * (converted to an array of 32 bytes), or 32 uniformly random bytes.
+   */
+  setMusig2NoncesHD(hdKeyPair: BIP32Interface, sessionId?: Buffer): void {
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:1042]:
```suggestion
  setMusig2NoncesHD(hdKeyPair: BIP32Interface, sessionId?: Buffer): this {
```

fluent interface

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 1 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 7]
```diff
@@ -391,22 +391,30 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+   * Otherwise, only derives from respective Bip32Derivation.
    *
+   * If no tapBip32Derivation or bip32Derivation is found, return the derived of the bip32Interface.
    * @param inputIndex
    * @param bip32
+   * @param inputType taproot or non-taproot
    * @private
    */
-  private getBip32PublicKeyForInput(inputIndex: number, bip32: BIP32Interface): Buffer | undefined {
+  private getSignerKeyPairForInput(
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 2 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 3 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 8]
```diff
@@ -971,15 +980,15 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint> = UtxoTransaction<bigin
...
+    const participantPubKey = participantPubKeys.find((pubKey) => pubKey.equals(signerKeyPair.publicKey));
 
     if (!Buffer.isBuffer(participantPubKey)) {
       throw new Error('participant plain pub key should match one bip32Derivation plain pub key');
     }
 
     const { hash } = this.getTaprootHashForSig(inputIndex);
     const pubNonce = this.nonceStore.createMusig2Nonce(
-      derivedWalletKey.privateKey,
+      signerKeyPair.privateKey,
```
[the remark this answers is not in the document]
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 5 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 6 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 4 above)
Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Done

**Review by OttoAllmendinger — APPROVED** [review]

**Pull request merged**